src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1425,6 +1449,121 @@ data LockingKind
       --
       -- @since 2.2.7
 
+newtype (ComparableLockStrength a) => StrongestLock a =
+  StrongestLock { getStrongestLock :: Maybe a}
+
+class ComparableLockStrength a where
+  -- | Create a lock that is at least as strong as both locks and at most as
+  -- strong as the stronger lock
+  -- Should satisfy:
+  --    If compare a b = GT
+  --        compareLockStrength (unionLocks a b) b = GT
+  --        compareLockStrength a (unionLocks a b) = EQ
+  --    If compare a b = EQ
+  --        compareLockStrength (unionLocks a b) b = EQ
+  --        compareLockStrength a (unionLocks a b) = EQ
+  --    If compare a b = LT
+  --        compareLockStrength (unionLocks a b) b = EQ
+  --        compareLockStrength a (unionLocks a b) = LT
+  unionLocks :: ComparableLockStrength a => a -> a -> a
+  compareLockStrength :: ComparableLockStrength a => a -> a -> Ordering
+
+instance ComparableLockStrength a => Semigroup (StrongestLock a) where
+  (<>) (StrongestLock Nothing) pright = pright
+  (<>) pleft (StrongestLock Nothing) = pleft
+  (<>) (StrongestLock (Just pright)) (StrongestLock (Just pleft)) =
+    case compareLockStrength pleft pright of
+      GT -> StrongestLock (Just pright)
+      EQ -> StrongestLock (Just $ unionLocks pright pleft)
+      LT -> StrongestLock (Just pleft)
+
+instance ComparableLockStrength a => Monoid (StrongestLock a) where
+  mempty = StrongestLock Nothing
+
+-- | Postgres specific locking, used only internally
+--
+-- @since 3.5.9.0
+data PostgresLockingKind =
+  PostgresLockingKind
+    {
+      postgresRowLevelLockStrength :: PostgresRowLevelLockStrength
+    , postgresLockingOfClause :: Maybe LockingOfClause
+    , postgresOnLockedBehavior :: OnLockedBehavior
+    }
+
+instance ComparableLockStrength PostgresLockingKind where
+  unionLocks pl1 pl2 =
+    let newOfClause = do
+          (Of e1) <- postgresLockingOfClause pl1
+          (Of e2) <- postgresLockingOfClause pl2
+          pure $ Of (e1 :& e2) in
+    case compareLockStrength pl1 pl2 of
+      GT -> pl1 { postgresLockingOfClause = newOfClause }
+      EQ -> pl1 { postgresLockingOfClause = newOfClause }
+      LT -> pl2 { postgresLockingOfClause = newOfClause }
+  compareLockStrength a b = compare (postgresRowLevelLockStrength a) (postgresRowLevelLockStrength b)
+
+-- Arranged in order of lock strength
+data PostgresRowLevelLockStrength =
+    PostgresForUpdate
+    | PostgresForShare
+  deriving (Ord, Eq)
+
+data LockingOfClause where
+  Of :: LockableEntity a => a -> LockingOfClause
```

Comment:
```suggestion
  LockingOfClause :: LockableEntity a => a -> LockingOfClause
```

`Of` is a pretty small word and it would be difficult to track what it's from - sharing the name will make it easier